### ISSUE TITLE
Master node wont rejoin etcd cluster after reboot

### ISSUE THREAD

**Issue by PhoenixMage** [k3s-io/k3s#2509]
**Environmental Info:**
K3s Version: 
k3s version v1.19.3+k3s2 (f8a4547b)

Node(s) CPU architecture, OS, and Version: 
3 nodes = 2 x arm64 and 1 x arm
k3os with manually upgraded k3s binary

Cluster Configuration:
3 masters in etcd HA configuration

**Describe the bug:**
After rebooting one of the arm64 masters it will not rejoin the cluster.

**Steps To Reproduce:**
Create 3 node HA etcd, reboot a node

**Expected behavior:**
Rebooted node should rejoin the cluster and sync with the quorum

**Actual behavior:**
k3s is in a crush loop and doesnt rejoin the cluster.

**Additional context / logs:**
The fatal message is below, not sure how to get more details from flannel in k3s.
FATA[2020-11-12T01:45:18.435959565Z] flannel exited: failed to acquire lease: nodes "rewind" is forbidden: not yet ready to handle request

Deleting the node from the cluster, then `sudo rm -rf /var/lib/rancher/k3s/server/db/etcd` and rejoining the node seems to be a work around.

**Comment by brandond** [comment]
Can you provide full k3s logs from the node in question on the startup where it exits with a fatal error?

**Comment by PhoenixMage** [comment]
Sure
[k3s error.log](https://github.com/rancher/k3s/files/5528017/k3s.error.log)

**Comment by brandond** [comment]
It would appear that this node is unable to respond to apiserver requests in a timely manner - likely it is either CPU or IO bound. Things are timing out after the 10 second internal maximum request duration, which causes components to fail and terminate the k3s process.

Note that restarting an existing node may put higher momentary demand on your system than joining a new one, as it has everything ready to start up all at once.

```
I1112 03:07:42.701903    4641 trace.go:205] Trace[1796645753]: "Create" url:/api/v1/namespaces/default/events,user-agent:k3s/v1.19.3+k3s2 (linux/arm64) kubernetes/f8a4547,client:127.0.0.1 (12-Nov-2020 03:07:32.695) (total time: 10005ms):
Trace[1796645753]: [10.005781913s] [10.005781913s] END
I1112 03:07:42.767729    4641 trace.go:205] Trace[1331455109]: "Create" url:/api/v1/nodes,user-agent:k3s/v1.19.3+k3s2 (linux/arm64) kubernetes/f8a4547,client:127.0.0.1 (12-Nov-2020 03:07:32.753) (total time: 10013ms):
Trace[1331455109]: [10.013590897s] [10.013590897s] END
E1112 03:07:42.770214    4641 kubelet_node_status.go:92] Unable to register node "rewind" with API server: nodes "rewind" is forbidden: not yet ready to handle request
E1112 03:07:42.776319    4641 controller.go:178] failed to update node lease, error: Put "https://127.0.0.1:6443/apis/coordination.k8s.io/v1/namespaces/kube-node-lease/leases/rewind?timeout=10s": context deadline exceeded
I1112 03:07:42.799649    4641 trace.go:205] Trace[743993935]: "Update" url:/apis/coordination.k8s.io/v1/namespaces/kube-node-lease/leases/rewind,user-agent:k3s/v1.19.3+k3s2 (linux/arm64) kubernetes/f8a4547,client:127.0.0.1 (12-Nov-2020 03:07:32.796) (total time: 10002ms):
Trace[743993935]: [10.002900612s] [10.002900612s] END
E1112 03:07:42.799411    4641 event.go:264] Server rejected event '&v1.Event{TypeMeta:v1.TypeMeta{Kind:"", APIVersion:""}, ObjectMeta:v1.ObjectMeta{Name:"rewind.1646a3d50311f60c", GenerateName:"", Namespace:"default", SelfLink:"", UID:"", ResourceVersion:"", Generation:0, CreationTimestamp:v1.Time{Time:time.Time{wall:0x0, ext:0, loc:(*time.Location)(nil)}}, DeletionTimestamp:(*v1.Time)(nil), DeletionGracePeriodSeconds:(*int64)(nil), Labels:map[string]string(nil), Annotations:map[string]string(nil), OwnerReferences:[]v1.OwnerReference(nil), Finalizers:[]string(nil), ClusterName:"", ManagedFields:[]v1.ManagedFieldsEntry(nil)}, InvolvedObject:v1.ObjectReference{Kind:"Node", Namespace:"", Name:"rewind", UID:"rewind", APIVersion:"", ResourceVersion:"", FieldPath:""}, Reason:"Starting", Message:"Starting kubelet.", Source:v1.EventSource{Component:"kubelet", Host:"rewind"}, FirstTimestamp:v1.Time{Time:time.Time{wall:0xbfe3479d1c8d6e0c, ext:23410817774, loc:(*time.Location)(0x65bc760)}}, LastTimestamp:v1.Time{Time:time.Time{wall:0xbfe3479d1c8d6e0c, ext:23410817774, loc:(*time.Location)(0x65bc760)}}, Count:1, Type:"Normal", EventTime:v1.MicroTime{Time:time.Time{wall:0x0, ext:0, loc:(*time.Location)(nil)}}, Series:(*v1.EventSeries)(nil), Action:"", Related:(*v1.ObjectReference)(nil), ReportingController:"", ReportingInstance:""}': 'events "rewind.1646a3d50311f60c" is forbidden: not yet ready to handle request' (will not retry!)
I1112 03:07:42.818792    4641 trace.go:205] Trace[450732774]: "GuaranteedUpdate etcd3" type:*coordination.Lease (12-Nov-2020 03:07:32.797) (total time: 10021ms):
Trace[450732774]: [10.021297721s] [10.021297721s] END
```

**Comment by PhoenixMage** [comment]
Its a Raspberry Pi 3b, is there a way to extend that timeout?

**Comment by brandond** [comment]
No, it's built in to Kubernetes. What sort of storage are you using? A faster SD card or moving /var/lib/rancher to an external disk might help.

**Comment by PhoenixMage** [comment]
Its a SanDisk Ultra 16Gb which I think is UHS-1. I wouldnt have thought it would be that constrained. My other 2 devices have eMMC on the board so that should be quikcer but I havent had a chance to test yet.

**Comment by brandond** [comment]
I generally go with UHS-II / U3 / Class 10. I'm not sure exactly what that translates to but it seems to work well enough.

**Comment by PhoenixMage** [comment]
Thanks, however I don't think Raspberry Pi supports UHS-II.

I will try overclocking the microsd controller and see if there is any improvement. The SanDisk Ultra cards I am using arent terrible... http://www.pidramble.com/wiki/benchmarks/microsd-cards#3-model-b

**Comment by brandond** [comment]
I haven't benchmarked mine, but they seem to work well enough. I have a cluster (2x Pi4B, 1x Pi3B) that are all using class 10 cards. That said I generally only use the Pi3 as an agent since it doesn't have enough memory to be a server and also run much useful workload.

**Comment by voipas** [comment]
Hey, any updates on this topic? As I am having identicalissue. I'm using Raspberry Pi 3b+ with Debian arm64 bit OS.

**Comment by luyfigue** [comment]
Just linking this to https://github.com/k3s-io/k3s/issues/2425

I had exactly the same problem after upgrading from 1.17.
As suggested, `--disable-cloud-controller` fixed the timeout at start up

**Comment by nirui** [comment]
Not sure if my issue was related.

But I got the same `flannel exited: failed to acquire lease: nodes "<Node Name>" is forbidden: not yet ready to handle request` error after a reboot, and then the cluster keeps crashing.

 I fixed the problem by disabling the builtin `flannel` and install my own instead. Basically, I changed it to following in my `k3s.service`:

    ExecStart=/usr/local/bin/k3s \
        server \
            '--flannel-backend=none' \
            '--disable-network-policy' \
            '--cluster-cidr=10.42.0.0/16' \

and then, when the cluster started up, install https://github.com/containernetworking/plugins/releases and run `kubectl apply -f https://raw.githubusercontent.com/coreos/flannel/master/Documentation/kube-flannel.yml`

I don't think this problem was related with the speed of the SD cards, because I used the same card, and the problem just gone after I disabled the built in `flannel`.

Maybe this will kick you smart & nice guys to look into this issue a little more? *blink* *blink* I mean...maybe the built in `flannel` should ... you know, just keep retrying instead of crash? Hehe :D
